### PR TITLE
Support Batches in ForwardAccumulator

### PR DESCRIPTION
**Pull request by abhichou4**
cc @allenlavoie

### CODE REVIEW

File: tensorflow/python/eager/pywrap_tfe_src.cc [hunk 1]
```diff
@@ -1535,6 +1535,10 @@ typedef tensorflow::eager::ForwardAccumulator<PyObject, PyBackwardFunction,
                                               PyTapeTensor>
     ForwardAccumulator;
 
+typedef tensorflow::eager::ForwardAccumulator<PyObject, PyBackwardFunction,
```
Comment (allenlavoie) [inline tensorflow/python/eager/pywrap_tfe_src.cc:1538]: ForwardBatchAccumulator?

Please add a test that would have caught this.

File: tensorflow/c/eager/tape.h [hunk 2]
```diff
@@ -1062,7 +1105,73 @@ Status ForwardAccumulator<Gradient, BackwardFunction, TapeTensor>::Accumulate(
...
+        in_grads.push_back(nullptr);
+      }
+    } else {
+      in_grads.push_back(current_grad->second);
+    }
+  }
+
+  std::vector<Gradient*> forward_grads;
+  if (forward_function == nullptr) {
+    // Raise apt error
```
Comment (allenlavoie) [inline tensorflow/c/eager/tape.h:1172]: Not sure what this means

File: tensorflow/c/eager/tape.h [hunk 3]
```diff
@@ -1062,7 +1105,73 @@ Status ForwardAccumulator<Gradient, BackwardFunction, TapeTensor>::Accumulate(
...
+        accumulated_gradients_[output_tensors[i].GetID()] = forward_grads[i];
+      }
+    }
+  }
+  return Status::OK();
+}
+
+template <typename Gradient, typename BackwardFunction, typename TapeTensor>
+Status
+ForwardBatchAccumulator<Gradient, BackwardFunction, TapeTensor>::Accumulate(
```
Comment (allenlavoie) [inline tensorflow/c/eager/tape.h:1131]: One thing that seems plausible is to just keep the boolean around, then we can add broadcasting for un-batched tangents in Python. Does that work, and do we need different C++ classes then?

File: tensorflow/python/tfe_wrapper.cc [hunk 4]
```diff
@@ -369,30 +369,28 @@ PYBIND11_MODULE(_pywrap_tfe, m) {
...
-      "TFE_NewContext",
-      [](const TFE_ContextOptions* opts) {
-        tensorflow::Safe_TF_StatusPtr status =
-            tensorflow::make_safe(TF_NewStatus());
-        TFE_Context* context = TFE_NewContext(opts, status.get());
-        tensorflow::MaybeRaiseRegisteredFromTFStatus(status.get());
-        return tensorflow::PyoOrThrow(tensorflow::OutputTFE_Context(context));
-      },
-      py::return_value_policy::reference);
+  m.def("TFE_NewContext",
```
Comment (allenlavoie) [inline tensorflow/python/tfe_wrapper.cc:372]: A minor nit, but it looks like your editor/code formatter is making lots of unrelated format changes. It makes the diff hard to interpret.

**Comment by gbaned** [comment]
@abhichou4  Can you please check @allenlavoie's comments and keep us posted ? Thanks!

File: tensorflow/python/tfe_wrapper.cc [hunk 5]
```diff
@@ -730,8 +727,9 @@ PYBIND11_MODULE(_pywrap_tfe, m) {
         });
 
   // TFE_Py_ForwardAccumulator logic.
-  m.def("TFE_Py_ForwardAccumulatorNew", []() {
-    return tensorflow::PyoOrThrow(TFE_Py_ForwardAccumulatorNew());
+  m.def("TFE_Py_ForwardAccumulatorNew", [](const py::handle& use_batch) {
```
[the remark this answers is not in the document]
Comment (abhichou4) [reply tensorflow/python/tfe_wrapper.cc]: Made change.

**Comment by abhichou4** [comment]
Adding tests for nested Accumulator/GradientTape and tf.function soon.

File: tensorflow/python/eager/forwardprop_test.py [hunk 6]
```diff
@@ -1,7 +1,4 @@
 # Copyright 2019 The TensorFlow Authors. All Rights Reserved.
-#
-# Licensed under the Apache License, Version 2.0 (the "License");
-# you may not use this file except in compliance with the License.
```
Comment (allenlavoie) [inline tensorflow/python/eager/forwardprop_test.py:4]: Why the license block change?

File: tensorflow/python/eager/forwardprop.py [hunk 7]
```diff
@@ -326,7 +326,7 @@ class ForwardAccumulator(object):
   <tf.Tensor: shape=(2,), dtype=float32, numpy=array([6., 0.], dtype=float32)>
   """
 
-  def __init__(self, primals, tangents):
+  def __init__(self, primals, tangents, use_batch=False):
```
Comment (allenlavoie) [inline tensorflow/python/eager/forwardprop.py:329]: Can we keep this out of the public API to start, e.g. a separate static method with this signature (_batch_accumulator)?

Once it's tested with tf.function/custom_gradient/nesting we can start with experimental_use_batch. Just a question of what's guaranteed about API stability.

**Review by allenlavoie — COMMENTED** [review]
I'd do those in another PR. If it just works it'll be a simple PR, but that's fine.

File: tensorflow/python/eager/forwardprop.py (hunk 7 above)
Comment (abhichou4) [reply tensorflow/python/eager/forwardprop.py:329]: That seems better. So instead of a `self` argument a function? Can we make it a `@classmethod`?

Comment (allenlavoie) [reply tensorflow/python/eager/forwardprop.py:329]: Sure, a new _private classmethod sounds reasonable to me for now.

File: tensorflow/python/eager/forwardprop.py [hunk 8]
```diff
@@ -439,3 +439,7 @@ def _fetch_jvp(tensor):
         return array_ops.zeros_like(tensor)
       return result
     return nest.map_structure(_fetch_jvp, primals)
+
+  @classmethod
+  def _batch_accumulator(cls, primals, unconnected_gradients=UnconnectedGradients.NONE):
```
Comment (allenlavoie) [inline tensorflow/python/eager/forwardprop.py:444]: I think using .jvp for fetching batch jvps is fine once batching is enabled, it's the constructor argument I'm concerned about. Can we take out the constructor argument and instead construct a batching ForwardAccumulator with a classmethod for now?

The issue is whether there's a public API for creating/using the batching mode. I don't think there should be until we know it composes with other common transformations.

File: tensorflow/python/eager/forwardprop.py [hunk 9]
```diff
@@ -439,3 +439,19 @@ def _fetch_jvp(tensor):
         return array_ops.zeros_like(tensor)
       return result
     return nest.map_structure(_fetch_jvp, primals)
+
+  @classmethod
+  def _batch_accumulator(cls, primals, tangents):
```
Comment (allenlavoie) [inline tensorflow/python/eager/forwardprop.py:444]: Can you add a one-line docstring?

File: tensorflow/c/eager/tape.h [hunk 10]
```diff
@@ -314,6 +315,9 @@ class ForwardAccumulator {
   // available in language bindings (e.g. Python).
   const VSpace<Gradient, BackwardFunction, TapeTensor>& vspace_;
 
+  //Decides if tangents are vectorized or not
+  bool use_batch;
```
Comment (allenlavoie) [inline tensorflow/c/eager/tape.h:319]: use_batch_ (trailing underscore): https://www.tensorflow.org/community/contribute/code_style#c_coding_style / https://google.github.io/styleguide/cppguide.html#Variable_Names

File: tensorflow/python/eager/forwardprop.py [hunk 11]
```diff
@@ -439,3 +439,19 @@ def _fetch_jvp(tensor):
...
+    acc = super(ForwardAccumulator, cls).__new__(cls, primals, tangents)
+    acc._recording = False
+    acc._accumulator = pywrap_tfe.TFE_Py_ForwardAccumulatorNew(True)
+    primal_ids = set()
+    for primal in nest.flatten(primals):
```
Comment (allenlavoie) [inline tensorflow/python/eager/forwardprop.py:449]: This might be a nice place to do some shape checking. Something like `tangent.shape.assert_is_compatible_with(tf.TensorShape(None) + primal.shape)`

**Review by allenlavoie — COMMENTED** [review]
Neat! Just a couple more nits.

**Review by allenlavoie — DISMISSED** [review]
This is great!

**Comment by allenlavoie** [comment]
Can you check the lint errors?

**Review by allenlavoie — APPROVED** [review]
